Initial support for extended integration test

mfojtik opened the pull request:
This is highly WIP atm. I have this here just to gather feedback and thoughts ;-)

mfojtik commented:
@bparees @csrwng @soltysh PTAL

To execute extended tests, you have to:

```
OS_TEST_TAGS="integration extended docker" ./hack/test-integration.sh -v=5
```

I don't like that I need to start the OpenShift server (all-in-one), since that breaks the parallelization of tests (I dunno if it is good idea to have multiple OpenShift server/etcd/... running simultaneously)...

mfojtik commented on test/integration/build_context_dir_test.go:
```diff
@@ -0,0 +1,75 @@
+	retries := 0
+
+	for {
+		time.Sleep(time.Duration(1 * time.Second))
```
FYI. this should be probably solved by using Watch() instead of polling, I just need to get familiar with the Watch() itself first...

csrwng commented:
> I don't like that I need to start the OpenShift server (all-in-one), since that breaks the parallelization of tests (I dunno if it is good idea to have multiple OpenShift server/etcd/... running simultaneously)...

Could we start the all-in-one server, set it up, then run all tests in parallel (each using its own namespace)?



> The second thing I don't like is that I can't verify the application is working without having Docker registry in place...

I was under the impression that what we wanted was a go version of the e2e tests with everything setup, including the registry. (so e2e extended posing as an integration test)

mfojtik commented:
@csrwng if extended tests will be a new set of tests, then I don't think I can re-use the code in integration... (dunno if I can just import the test_server.go ;-)

csrwng commented:
That's what I thought we were going to do :-)

mfojtik commented:
@csrwng I can still move this into its own package :-) (this is the feedback I was hoping for :-)

csrwng commented:
> @csrwng I can still move this into its own package :-) (this is the feedback I was hoping for :-)

You have my vote

mfojtik commented:
@csrwng so the command to launch the tests is now:

```
OS_TEST_PACKAGE="test/extended" OS_TEST_TAGS="integration extended docker" ./hack/test-integration.sh
```

soltysh commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,76 @@
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/test/integration"
```
Instead of having those bits and pieces here and there, let's create a `test/util` package and put there all the utility stuff such as `RequireDocker`, `RequireEtcd`, etc.

soltysh commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,76 @@
+
+func TestExtendedBuildContextDir(t *testing.T) {
+	integration.DeleteAllEtcdKeys()
+	openshift, _ := integration.StartTestAllInOne()
```
Yet again candidate for `test/util` package

mfojtik replied:
yeah, I hate to type integration. everytime I want to call this... also funny fact... we are not using StartServer anywhere in integration....

csrwng commented:
> @csrwng so the command to launch the tests is now:

we should add a hack/test-extended.sh to run it

soltysh commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,76 @@
+// +build integration,extended,!no-etcd,!no-docker,docker
```
I think we are overusing those, especially in that we do copy&paste for other integration tests. I think it would make sense to clear this out doing this. Since for this test you are requiring etcd and docker this line should state:
```
// +build integration,extended,etcd,docker
```
IMHO

mfojtik replied:
there should be no tags at all here...

soltysh replied:
I think you may want to leave `integration,extended` since you're passing those with `OS_TEST_TAGS`

bparees replied:
should it just be extended?  these tests aren't run when hack/test-integration.sh is run (because they're in a different package), so tagging them as integration is a bit confusing.

soltysh commented:
:sunflower: for the idea itself

soltysh commented:
> we should add a hack/test-extended.sh to run it

+1

bparees commented:
I think it's reasonable to start the registry service as part of the basic integration test environment setup.

and i agree w/ @csrwng that we should be able to run in parallel by using appropriate namespaces (though that probably means we need some more generic framework that will dynamically generate unique namespaces and substitute it into templates/etc, otherwise we're relying on everyone to create their tests using unique namespaces).

also currently namespaces don't work for services...upstream issue still pending.

bparees commented:
might be useful to have a soft-reset for openshift... ie "delete all pods/etc, clean up etcd (maybe just cleaning up etcd is sufficient), stop all docker containers"

might be faster than starting openshift again for each test.

though given that you'd need to purge/restart the registry too i guess it probably wouldn't save much time.....i just worry about how long each integration test is going to take if we have to start openshift and wait for the registry to deploy....that's like 60s just in startup time per test.  I know part of the point of this is to have a place for longer tests to live, but still.

soltysh commented:
I guess then it'll be simpler and faster at the same time, to generate new namespace for each test. Eg. we could be generating env var called `TEST_NAMESPACE` and all tests would be then consuming that.

soltysh commented on hack/test-integration.sh:
```diff
@@ -37,21 +37,25 @@ mkdir -p "${testdir}"
 # build the test executable (cgo must be disabled to have the symbol table available)
 pushd "${testdir}" 2>&1 >/dev/null
-CGO_ENABLED=0 go test -c -tags="${tags}" "${OS_GO_PACKAGE}/${package}"
+CGO_ENABLED=0 go test -v -c -tags="${tags}" "${OS_GO_PACKAGE}/${package}"
```
Can't we have that verbosity connected with `VERBOSE` env var?

smarterclayton replied:
Why would you need the build to be verbose?

soltysh commented on hack/test-integration.sh:
```diff
@@ -37,21 +37,25 @@ mkdir -p "${testdir}"
-    echo ok $1
-    exit 0
+  if [[ ! -z "${VERBOSE-}" ]]; then
+    ${testexec} -logtostderr=true -test.v=true -v=5 -test.run="^$1$" "${@:2}"
```
Can't we use `VERBOSE` to set actual verbosity log instead of hardcoding 5 here.

mfojtik replied:
we can :-) good catch

mfojtik commented:
@bparees @csrwng 

I updated this:

* OpenShift server is now started by test-extended.sh
* GetClient() method was added into server.go so you can obtain an osclient you use in your tests
* More smaller refactorings and tweaks to have these tests running consistently

PTAL

mfojtik commented:
FYI, now you execute the tests using:

```
./hack/test-extended.sh
```

(you can optionally set VERBOSE=1 before you run that command to get more verbose output...)

csrwng commented on hack/test-extended.sh:
```diff
@@ -0,0 +1,65 @@
+export OS_EXTENDED_MASTER=127.0.0.1:54199
+export OS_EXTENDED_ASSETS=127.0.0.1:54299
+
+export BASETMPDIR="/tmp/openshift-extended"
```
Using ${TMPDIR:-/tmp} instead of /tmp would be useful in case you want to use a different temp directory

csrwng commented on test/util/docker.go:
```diff
@@ -0,0 +1,39 @@
+	client, err := NewDockerClient()
+
+	if err != nil {
+		glog.Fatalf("unable to create docker client for integration testing: %v", err)
```
s/integration//

csrwng commented on test/util/docker.go:
```diff
@@ -0,0 +1,39 @@
+	_, err = client.ListImages(dockerClient.ListImagesOptions{All: false})
+
+	if err != nil {
+		glog.Fatalf("unable to create docker client for integration testing: %v", err)
```
s/integration//

csrwng commented on test/util/etcd.go:
```diff
@@ -0,0 +1,66 @@
+// RequireEtcd verifies if the etcd is running and accessible for testing
+func RequireEtcd() {
+	if _, err := NewEtcdClient().Get("/", false, false); err != nil {
+		glog.Fatalf("unable to connect to etcd for integration testing: %v", err)
```
s/integration//

csrwng commented on hack/test-extended.sh:
```diff
@@ -0,0 +1,65 @@
+
+# Start OpenShift sever that will be common for 
+echo "[INFO][$(date)] Starting OpenShift test server [${OS_EXTENDED_MASTER}]..."
+start_server
```
It'd also be useful to instantiate the docker registry after starting the server.

mfojtik replied:
+1 does the registry creation command offer 'watch' ?

csrwng replied:
It completes when it has created the artifacts, but I don't think it ensures that they are running

bparees replied:
it does not, borrow the loop from test-end-to-end.sh.

csrwng commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+func mockBuildWithContextDir() *buildapi.Build {
```
naming? this isn't really a mock, it's an actual build.

csrwng commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,68 @@
+func TestBuildContextDir(t *testing.T) {
+	build := mockBuildWithContextDir()
+	client := util.GetClient()
+	if _, err := client.Builds("contextdir").Create(build); err != nil {
```
maybe a random namespace would be better than fixed "contextdir". That way in theory you could run this test multiple times in parallel if that's what you want to do.

bparees replied:
+1.  contextdir+timestamp(milliseconds)

mfojtik replied:
I will make this helper function

csrwng commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,68 @@
+	retries := 0
+
+	for {
+		time.Sleep(time.Duration(1 * time.Second))
```
Was the thinking to switch these to use the watch api at some point?

mfojtik replied:
Yes, there was a TODO before and I can't remember what led me to hide it :-) I will rework this to use watch.

csrwng commented on hack/test-extended.sh:
```diff
@@ -0,0 +1,65 @@
+source ${OS_ROOT}/hack/util.sh
+source ${OS_ROOT}/hack/common.sh
+
+# TODO: Randomize these ports
```
Is the reason for using random ports that you could run the tests in parallel on a single machine?

mfojtik replied:
Yes, that was my thinking... We have random ports when you start the server using utils/server.go (StartAllInOneServer?)... Problem is that having random ports also means we need to have bash function that will scan if the ports is free... I don't want to add this now... I think I will add more reasoning to this comment

bparees replied:
pick a port based on the PID.

technically you should still collide w/ some other process, but at least you won't collide between test invocations.

csrwng replied:
I would think you'd also want a random etcd port

mfojtik replied:
@bparees @csrwng I'm thinking about making a small Go helper that returns free port (same as we're doing here: https://github.com/openshift/origin/blob/master/test/integration/server_test.go#L44)

(go run test/util/helpers/address.go => 127.0.0.1:RANDOM)

bparees commented on hack/test-extended.sh:
```diff
@@ -0,0 +1,65 @@
+    --nodes=127.0.0.1 \
+    --hostname="127.0.0.1" \
+    --cert-dir="${CERT_DIR}" \
+    --loglevel=3 &> ${BASETMPDIR}/server.log &
```
--latest-images?

bparees commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,68 @@
+		switch current.Status {
+		case buildapi.BuildStatusFailed, buildapi.BuildStatusError:
+			t.Fatalf("Unexpected failure of build: %v", current)
+		case buildapi.BuildStatusComplete:
```
the build completing doesn't prove that you actually built the app using the correct contextDir or that the output was valid/correct.  needs more validation imho.

bparees commented on test/extended/build_docker_test.go:
```diff
@@ -0,0 +1,68 @@
+		switch current.Status {
+		case buildapi.BuildStatusFailed, buildapi.BuildStatusError:
+			t.Fatalf("Unexpected failure of build: %v", current)
+		case buildapi.BuildStatusComplete:
```
again completion!=success.  I think we need a mock sti + mock custom builder that 1) runs super (trivial assemble script) and 2) produces an output that can easily be validated (less sure how we can easily do that w/o having to startup up the resulting container, but even if we have to start the output container, we can start it via docker directly rather than via k8s, which saves some time...no scheduler/deployer/etc)

smarterclayton commented on hack/test-go.sh:
```diff
@@ -24,6 +24,7 @@ find_test_dirs() {
         -o -wholename '*/Godeps/*' \
         -o -wholename '*/_output/*' \
         -o -wholename './.git' \
+        -o -wholename './test/extended/*' \
```
This shouldn't to be necessary - your test cases are already guarded with build tags

smarterclayton commented on hack/test-integration.sh:
```diff
@@ -37,21 +37,25 @@ mkdir -p "${testdir}"
 popd 2>&1 >/dev/null
 
-start_etcd
+[ "${ETCD_STARTED-}" != "1" ] && start_etcd
```
Move this inside start_etcd

mfojtik commented:
@bparees @csrwng FYI, I addressed all comments (thanks!) and currently working on improving the test cases (push to ImageRepo, watches, etc...) Right now I have fully working OpenShift server, Registry installed, etc

mfojtik commented:
FYI: This will need squash, just want to track what I'm doing ;-)

mfojtik commented on hack/test-extended.sh:
```diff
@@ -0,0 +1,101 @@
+export OS_TEST_TAGS="extended"
+export OS_TEST_NAMESPACE="extended"
+
+[ ! -z "${RUN_SERVER-}" ] && wait;exit
```
FYI, ignore this for now ;-) I'm using this to keep server running and don't run any test cases, so I can test helper methods/etc without restarting the server....

mfojtik commented:
... another rebase hell ;-(

soltysh commented on test/extended/build_context_dir_test.go:
```diff
@@ -0,0 +1,69 @@
+		// the build is running, complete, or has failed
+		if build.Name == newBuild.Name {
+			switch build.Status {
+			case buildapi.BuildStatusFailed, buildapi.BuildStatusError:
```
I'd add `BuildStatusCancelled` here as well, as it's also an end status of a build.

mfojtik replied:
@soltysh unlikely in the test run (I mean you have to Cancel the build by manual intervention)

soltysh commented on test/extended/build_docker_test.go:
```diff
@@ -0,0 +1,61 @@
+		// the build is running, complete, or has failed
+		if build.Name == newBuild.Name {
+			switch build.Status {
+			case buildapi.BuildStatusFailed, buildapi.BuildStatusError:
```
I'd add `BuildStatusCancelled` here as well, as it's also an end status of a build.

mfojtik commented:
[test]

mfojtik commented:
@deads2k can you please check the latest failure? I think it has something to do with Policies vs Namespaces but I can't figure out what broke there....

mfojtik commented:
@deads2k this might be related:

```
I0319 11:01:44.033426    7246 master.go:442] No master policy found.  Creating bootstrap policy based on: /tmp/openshift-integration/policy/policy.json
E0319 11:01:44.265429    7246 master.go:445] Error creating bootstrap policy: policyBinding "openshift" not found

```

mfojtik commented:
Also I wonder why TestDNS requires Docker:

```
W0319 11:32:07.796088    7921 node.go:96] Error running 'chcon' to set the kubelet volume root directory SELinux context: exit status 1
F0319 11:32:07.796252    7921 node.go:61] ERROR: Docker could not be reached at unix:///var/run/docker.sock.  Docker must be installed and running to start containers.
```

mfojtik commented:
@csrwng @bparees @soltysh PTAL

I would really like to have this merged ASAP as this is getting broken everyday (as somebody push something into integration tests). I can fix small issues as follow up.

csrwng commented:
LGTM but it looks like integration tests are still failing on go 1.4

mfojtik commented:
@csrwng yeah, I saw this failure on other PR as well and @deads2k is looking into it... It does not seem that my changes made it worse/better.

mfojtik commented:
[merge] (will fix all issues as follow up)

mfojtik commented:
[merge] again.. seems like random failure
